src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1914 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "bus-util.h"
+#include "hashmap.h"
+#include "log.h"
+#include "sd-bus.h"
+
+#include "build-path.h"
+#include "bus-error.h"
+#include "bus-get-properties.h"
+#include "bus-label.h"
+#include "bus-log-control-api.h"
+#include "bus-polkit.h"
+#include "common-signal.h"
+#include "discover-image.h"
+#include "env-util.h"
+#include "event-util.h"
+#include "escape.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "main-func.h"
+#include "memfd-util.h"
+#include "mkdir-label.h"
+#include "os-util.h"
+#include "process-util.h"
+#include "sd-json.h"
+#include "service-util.h"
+#include "signal-util.h"
+#include "socket-util.h"
+#include "string-table.h"
+
+typedef struct Manager {
+        sd_event *event;
+        sd_bus *bus;
+
+        Hashmap *targets;
+
+        uint64_t current_job_id;
+        Hashmap *jobs;
+
+        Hashmap *polkit_registry;
+
+        sd_event_source *notify_event;
+} Manager;
+
+/* Forward declare so that jobs can call it on exit */
+static void manager_became_idle(Manager *m);
+
+typedef enum TargetClass {
+        /* These should try to match ImageClass from src/basic/os-util.h */
+        TARGET_MACHINE = IMAGE_MACHINE,
+        TARGET_PORTABLE = IMAGE_PORTABLE,
+        TARGET_SYSEXT = IMAGE_SYSEXT,
+        TARGET_CONFEXT = IMAGE_CONFEXT,
+        _TARGET_CLASS_IS_IMAGE_CLASS_MAX,
+
+        /* sysupdate-specific classes */
+        TARGET_HOST = _TARGET_CLASS_IS_IMAGE_CLASS_MAX,
+        TARGET_COMPONENT,
+
+        _TARGET_CLASS_MAX,
+        _TARGET_CLASS_INVALID = -EINVAL,
+} TargetClass;
+
+/* Let's ensure when the number of classes is updated things are updated here too */
+_Pragma("GCC diagnostic push");
+_Pragma("GCC diagnostic ignored \"-Wenum-compare\"")
+assert_cc(_IMAGE_CLASS_MAX == _TARGET_CLASS_IS_IMAGE_CLASS_MAX);
```

Comment:
i'd just cast to (int) here instead of doing the pragma thing.

i.e.:

```c
assert_cc((int) _IMAGE__CLASS_MAX == (int) _TARGET_CLASS_IS_IMAGE_CLASS_MAX);
```

Reply:
(also, if we really need to do things like this, we prefer doing this via the DISABLE_WARNING_xx macros)